E2E: add coverage for user upgrade from Free to Pro plan.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR adds coverage for the scenario where a user of an existing Free site upgrades to the Pro plan.

Key changes:
- addition of methods in RestAPIClient to support creation of a site;
- reworking the existing spec `Plans: Add Upgrade to Cart` to a new spec `Plans: Upgrade exising WordPress.com Free site to WordPress.com Pro`;

#### Testing instructions

Ensure the following:
  - [x] Pre-Release E2E

Related to https://github.com/Automattic/wp-calypso/issues/62492.
Depends on https://github.com/Automattic/wp-calypso/pull/63981.

dpasque commented on test/e2e/specs/plans/plans__existing-free-site-upgrade.ts:
```diff
@@ -0,0 +1,150 @@
+				throw new Error( `Failed to create new site via REST API.\nHTTP response: ${ response }` );
+			}
+			newSiteDetails = response;
+			siteCreatedFlag = response.body.success;
```
I missed this in the other PRs (my apologies), but I think it might be more clear to update the typing for `NewSiteResponse` to be just the _body_ of the response, not the wrapper Playwright object that also includes status code.

Across all the Playwright response objects, the only thing that's going to differentiate meaningfully is the bodies. If we want to type the various responses, we could do a generic type like...

```typescript
interface Response< Body > {
   code: number;
   body: Body
}
```

Playwright might actually support that already!

I think this would also clean up a lot of the calling code so you don't have to keep differentiating from the wrapper object vs the more meaningful parsed body content.

worldomonation replied:
In the end this PR became really stale and required extensive rebase and rework. As part of that, I ended up merging the `NewSiteResponse` types used between the POMs and the API client into one; that means also removing the `body/code` part and focusing only on the metadata that matters.

worldomonation commented:
Needs a rebase but otherwise good to review.

dpasque commented on test/e2e/specs/plans/plans__existing-free-site-upgrade.ts:
```diff
@@ -0,0 +1,150 @@
+		} );
+
+		// Pre-hook to authenticate as user
+		beforeAll( async function () {
```
This is stylistic and optional, but having multiple `beforeAll` hooks might confuse some folks, and isn't really a pattern I've ever seen used JS/TS testing frameworks. I think it might be more clear if there's only one hook of a given type per stack level.

To get around having that be a really long hook, you could define some async functions inline and then just call them in the single beforeAll hook. That would make it clear the order of distinct steps. Like...

```typescript
const createFreeSite = async () => {} // those steps
const authenticate = async () => {}

beforeAll( async () => {
    await createFreeSite();
    await authenticate():
} );
```

worldomonation replied:
I've merged the `beforeAll` hooks into one now.

dpasque commented on test/e2e/specs/plans/plans__existing-free-site-upgrade.ts:
```diff
@@ -0,0 +1,150 @@
+				title: blogName,
+			} );
+
+			console.log( response );
+			console.log(
+				DataHelper.getCalypsoURL( `/home/${ newSiteDetails.body.blog_details.site_slug }` )
+			);
```
I figure you'll remove these before publishing, but just an FYI that the debug code is still in here 😄

worldomonation replied:
Thanks, removed!

dpasque reviewed (approved):
Overall, I really like the test design here -- looks great! Using the rest API is really slick. And the test spec reads so cleanly and clearly! 🎉 

I left a few comments, and I know there's some rebasing and TS failures to address, but I don't have any concerns with this and so will approve so you can merge as needed in the future! 😄

worldomonation commented:
On hold due to 773-gh-Automattic/martech.

worldomonation committed:
Start work on WordPress.com Free to Pro

Add `CheckoutThankYouPage`.

worldomonation committed:
Implement RestAPIClient.createSite method.

worldomonation committed:
Implement new spec "Plans: Upgrade exising WordPress.com Free site to WordPress.com Pro"

worldomonation committed:
Update RestAPIClient:
- add call to the `getBearerToken` when generating the authorization header.
- await on `response.json()` in `sendRequest`.
- unify on the `NewSiteResponse` type; previously the version used in POM and the RestAPIClient differed.

Split RestAPIClient tests into two.
- add tests for `createSite`.

worldomonation committed:
Revise the Plans: Upgrade exising WordPress.com Free site to WordPress.com Pro spec.
- merge the beforeAll hooks into one.
- remove extraneous calls.

worldomonation committed:
Plans: Upgrade exising WordPress.com Free site to WordPress.com Pro
- remove extrneous test step that was failing on mobile.

RestAPIClient unit tests:
- fix mocked data values

worldomonation committed:
Remove extrneous step from spec.

youbrokesomething commented:
@worldomonation Should this PR include turning the other new plan tests back on that were disabled in [64330](https://github.com/Automattic/wp-calypso/pull/64330) or should that be a different PR?

worldomonation commented:
> @worldomonation Should this PR include turning the other new plan tests back on that were disabled in [64330](https://github.com/Automattic/wp-calypso/pull/64330) or should that be a different PR?

I'd say we should turn it back on as part of a separate PR. Good call to do so, I had forgotten that we disabled those specs.
